kernel/mempool.c

```diff
@@ -405,4 +405,17 @@ void k_free(void *ptr)
 		k_mem_pool_free(ptr);
 	}
 }
+
+void *k_calloc(size_t nmemb, size_t size)
+{
+	void *ret;
+	size_t bounds;
+
+	bounds = nmemb * size;
```

Comment:
Need to verify if this is going to overflow.

Reply:
do we?
In Zephyr we generally do not do argument checking.
If a user passes in insane values, garbage in garbage out IMO

Reply:
An assertion wouldn't hurt.

Reply:
No opinion on the error checking personally.  All but certain the C11 & POSIX definitions allow calloc() to be undefined on overflow.

But as for pedantry, you need to align size for the platform.  If the user passes a "size" value that is not naturally aligned, then the individual members may be misaligned, and IIRC the standards *do* guarantee that memory blocks are aligned to the minimum the platform requires (0 on x86, 4 most other 32 bit platforms, etc...).

Basically, if you calculate a size for a block that is, say, 9 bytes, with a struct header and then a 5-byte string or whatever, and allocate an array of them here, the second one will be misaligned and a pointer read from that struct may cause an exception at runtime.  That's illegal.  I think.

Just a "size = (size + 3) & ~3" is enough for all our platforms I think.

Reply:
> you need to align size for the platform

> "size = (size + 3) & ~3" is enough for all our platforms I think.

I think this is wrong. This will clearly not work for creating an array of an integral type.
We do not expect calloc(30, sizeof(char)) to create a character array to return a blob of memory 120 bytes in size!
the second argument to calloc() should be the return value of some sizeof() call.
When creating an array of some struct, the sizeof() call on that struct will have the padding already. We should not be messing with the size argument like you propose.

As far as I can tell, the only alignment we need to worry about is the alignment of the base of the array.

Reply:
Good point on the alignment of smaller structs.  Pretty sure that somewhere in the spec, though, calloc() is required to return a natural alignment for every element (so if size==1 you can return a byte-aligned element, but not if size==3).  The glibc man page is silent on that though.  I'm fine with it as-is, I guess.

Reply:
This is what newlib does:

https://sourceware.org/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=newlib/libc/stdlib/mallocr.c;h=ecc445f3d36365a4840e31c737db5018ddba42e9;hb=refs/heads/master

This is what glibc does:

https://sourceware.org/git/?p=glibc.git;a=blob;f=malloc/malloc.c;h=f94d51cca1b35952b15926883bb0256229ffb1fe;hb=HEAD

In both cases they are just multiplying the element size by the number of elements to obtain the total memory size

Reply:
Actually, no, glibc does [check for overflow before relaying that to malloc()](https://sourceware.org/git/?p=glibc.git;a=blob;f=malloc/malloc.c;h=f94d51cca1b35952b15926883bb0256229ffb1fe;hb=HEAD#l3364).  It just doesn't use the GCC builtin for that, though.  [OpenBSD also does this](http://bxr.su/OpenBSD/lib/libc/stdlib/malloc.c#1709), as does [musl libc](https://git.musl-libc.org/cgit/musl/tree/src/malloc/calloc.c).  Newlib has two mallocs, and neither of them will perform any check.

In fact, this is [highlighted by the secure coding guidelines that CERT puts out](https://www.securecoding.cert.org/confluence/display/cplusplus/MEM07-CPP.+Ensure+that+the+arguments+to+calloc%28%29%2C+when+multiplied%2C+can+be+represented+as+a+size_t), because this can lead to many vulnerabilities.

I'd add an __ASSERT(). It's free on release builds and documents the invariant.

Reply:
@lpereira sure I'll work up a PR today to add an assertion